tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,672 @@
+    permits: AtomicUsize,
+}
+
+struct Waitlist {
+    queue: LinkedList<Waiter>,
+    closed: bool,
+}
+
+/// Error returned by `Permit::try_acquire`.
+#[derive(Debug)]
+pub(crate) enum TryAcquireError {
+    Closed,
+    NoPermits,
+}
+/// Error returned by `Permit::poll_acquire`.
+#[derive(Debug)]
+pub(crate) struct AcquireError(());
+
+#[derive(Debug)]
+pub(crate) struct Permit {
+    state: PermitState,
+}
+
+pub(crate) struct Acquire<'a> {
+    node: Waiter,
+    semaphore: &'a Semaphore,
+    permit: &'a mut Permit,
+    num_permits: u16,
+}
+
+/// Permit state
+#[derive(Debug, Copy, Clone)]
+enum PermitState {
+    /// Currently waiting for permits to be made available and assigned to the
+    /// waiter.
+    Waiting(u16),
+
+    /// The number of acquired permits
+    Acquired(u16),
+}
+
+/// An entry in the wait queue.
+struct Waiter {
+    /// The current state of the waiter. 
+    /// 
+    /// This is either the number of remaining permits required by
+    /// the waiter, or a flag indicating that the waiter is not yet queued.
+    state: AtomicUsize,
+
+    /// The waker to notify the task awaiting permits.
+    /// 
+    /// # Safety
+    /// 
+    /// This may only be accessed while the wait queue is locked.
+    /// XXX: it would be nice if we could enforce this...
+    waker: CausalCell<Option<Waker>>,
+
+    /// Intrusive linked-list pointers.
+    /// 
+    /// # Safety
+    /// 
+    /// This may only be accessed while the wait queue is locked.
+    pointers: linked_list::Pointers<Waiter>,
+
+    /// Should not be `Unpin`.
+    _p: PhantomPinned,
+}
+
+const CLOSED: usize = 1 << 17;
+
+fn notify_all(list: &mut LinkedList<Waiter>) {
+    while let Some(waiter) = list.pop_back() {
+        let waker = unsafe { 
+            waiter.as_ref().waker.with_mut(|waker| (*waker).take())
+        };
+
+        waker.expect("if a node is in the wait list, it must have a waker").wake();
+    }
+}
+
+impl Semaphore {
+    /// Creates a new semaphore with the initial number of permits
+    pub(crate) fn new(permits: usize) -> Self {
+        assert!(permits <= std::u16::MAX as usize);
+        Self {
+            permits: AtomicUsize::new(permits),
+            waiters: Mutex::new(Waitlist {
+                queue: LinkedList::new(),
+                closed: false,
+            }),
+        }
+    }
+
+    /// Returns the current number of available permits
+    pub(crate) fn available_permits(&self) -> usize {
+        self.permits.load(SeqCst) & std::u16::MAX as usize
+    }
+
+
+    /// Adds `n` new permits to the semaphore.
+    pub(crate) fn add_permits(&self, added: usize) {
+        if added == 0 {
+            return;
+        }
+
+        // Assign permits to the wait queue, returning a list containing all the
+        // waiters at the back of the queue that received enough permits to wake
+        // up.
+        let mut notified = self.add_permits_locked(added, self.waiters.lock().unwrap());
+        
```

Comment:
We shouldn't be waking inside the lock. `add_permits_locked` simply cuts off the tail of the queue that has received enough permits to be notified, returning a new list. We then drain _that_ list and notify, outside of the lock. Since `add_permits_locked` takes the `MutexGuard` by value, the lock should be released when that function returns.

Should maybe add a comment explaining this.

Reply:
Added some comments in a82c043f85e5d0a3fd67276e887be29466173254 which should make this clearer!